drivers/mil_passive_sonar/multilateration/multilateration.py

```diff
@@ -1,73 +1,184 @@
-    arrival to the individual sensors. c is the wave speed in the medium of operation.
-    Units:
-        Hydrohone coordinates are expected in millimeters, pulse location will be given in millimeters.
-        Timestamps are expected in microseconds. c is expected in millimeters per microsecond
-    Note:
-        hydrophone locations should be the dict returned by rospy.get_param('~/<node name>/hydrophones
+    This class uses difference in time of arrival (DTOA) measurements to calculate the relative
+    heading or postion of a signal source with respect to a receiver array.
+
+    There as of now, there are 3 different solvers that you can use. You can select using the 'method'
+    argument.
+    bancroft - analytical solution, fast but not very numerically stable
+    LS - optimization routine, fast and robust to noise
+    LS1 - optimization routine, slow, but may be more accurate
+
+    If the distance between receivers is orders of magnitude below the distance between the source
+    and the array, the resulting position estimate should be in the right heading direction but with
+    unreliable range. Otherwise, it may be possible to fully estimate the relative position of the
+    pinger.
+ 
+    Note: All distances are in millimeters and all times in microseconds
     '''
-    def __init__(self, hydrophone_locations, c, method):  # speed in millimeters/microsecond
-        self.hydrophone_locations = []
-        for key in hydrophone_locations:
-            sensor_location = np.array([hydrophone_locations[key]['x'], hydrophone_locations[key]['y'], hydrophone_locations[key]['z']])
-            self.hydrophone_locations += [sensor_location]
-        self.pairs = list(combinations(range(len(hydrophone_locations)),2))
+    def __init__(self, receiver_locations, c, method):  # speed in millimeters/microsecond
+        '''
+        receiver_locations - N x 3 numpy array with the locations of receivers in units of millimeters.
+            The location of the reference hydrophone is considered the origin of the receiver array's
+            reference frame and should not be included
+        c - speed at whic the pulse propagates through the medium (millimeters / microsecond)
+        method - kind of solver used to estimate the position or heading to the pulse
+        '''
+        self.receiver_locations = receiver_locations
+        self.n = len(receiver_locations)
+        self.pairs = list(combinations(range(self.n), 2))
         self.c = c
         self.method = method
+        self.solvers = {'bancroft' : self.estimate_pos_bancroft,
+                        'LS'       : lambda dtoa: self.estimate_pos_LS(dtoa, self.cost_LS),
+                        'LS1'      : lambda dtoa: self.estimate_pos_LS1(dtoa, self.cost_LS)}
         print "\x1b[32mSpeed of Sound (c):", self.c, "millimeter/microsecond\x1b[0m"
 
-    def getPulseLocation(self, timestamps, method=None):
+    def getPulseLocation(self, dtoa, method=None):
```

Comment:
Should_be_consistent_with_function_naming_scheme
Also should this always return something, you could get keyboard interrupted and not return anything (though anytime you get keyboard interrupted you probably want to stop the program anyway)